uast/ann/ann.go

```diff
@@ -14,9 +17,43 @@ const (
 	descendantOrSelf
 )
 
-// Predicate is a function that takes a node and returns a boolean. It
-// is provided as a convenient way of defining a PathMatcher.
-type Predicate func(n *uast.Node) bool
+// String returns the XPath (XML Path Language) string representation of
+// an axis.
+func (a axis) String() string {
+	switch a {
+	case self:
+		return "self"
+	case child:
+		return "child"
+	case descendant:
+		return "descendant"
+	case descendantOrSelf:
+		return "descendant-or-self"
+	default:
+		panic(fmt.Sprintf("unknown axis: %q", a))
+	}
+}
+
+// Predicate is the interface that wraps boolean tests for uast.Nodes.
+//
+// The Eval function evaluates the test over a node and returns its
+// boolean output.  The String function returns  a description of the
+// predicate in a syntax similar to XPath.
+type Predicate interface {
+	fmt.Stringer
+	Eval(n *uast.Node) bool
+}
+
+// predicate implements Predicate by using desc as the XPath
+// representation of the predicate.
+type predicate struct {
```

Comment:
I don't understand why Predicate and Action interfaces and related types include the XPath representation (stringer interface / desc field), instead of handling this XPath representation independently. Handling it in a decoupled way wouldn't require changes to existing predicates/actions and would make adding representations easier.

That being said, I don't see an easy way to implement it in a decoupled way, considering that rules are constructed through function calls, so forget about it.